File: modules/motocortex-obd/ios/MotoCortexOBDModule.swift
import ExpoModulesCore

public class MotoCortexOBDModule: Module {
  private let sessionManager = OBDSessionManager()
  // Direct Data Buffer for JSI Zero-Allocation Ring Buffer mapping
  private var telemetryBuffer = Data(count: 64)

  public func definition() -> ModuleDefinition {
    Name("MotoCortexOBDModule")

    Events("onConnectionStateChanged")

    OnCreate {
      self.sessionManager.setOnStateChangedListener { [weak self] state in
        self?.sendEvent("onConnectionStateChanged", [
          "state": state
        ])
      }

      self.sessionManager.setTelemetryListener { [weak self] data in
        self?.parseAndMutateTelemetry(data)
      }
    }

    AsyncFunction("connectDeviceAsync") { (type: String, target: String) async -> Bool in
      return await withCheckedContinuation { continuation in
        self.sessionManager.connect(type: type, target: target) { success in
          continuation.resume(returning: success)
        }
      }
    }

    AsyncFunction("disconnectDeviceAsync") {
      self.sessionManager.disconnect()
    }

    AsyncFunction("writeCommandAsync") { (command: String) async throws -> String in
      guard let q = self.sessionManager.queue else {
        throw NSError(domain: "OBD", code: -1, userInfo: [NSLocalizedDescriptionKey: "OBD Module: Device not connected"])
      }
      return try await withCheckedThrowingContinuation { continuation in
        q.enqueue(commandString: command, timeoutMs: 1000) { result in
          switch result {
          case .success(let response):
            continuation.resume(returning: response)
          case .failure(let error):
            continuation.resume(throwing: error)
          }
        }
      }
    }

    Function("abortPendingCommands") {
      self.sessionManager.queue?.abortPendingCommands()
    }

    /**
     * Returns a direct JSI buffer pointer mapping (maps to ArrayBuffer in JS).
     */
    Function("getTelemetryBuffer") { () -> Data in
      return self.telemetryBuffer
    }
  }

  private func parseAndMutateTelemetry(_ data: String) {
    let cleanData = data.replacingOccurrences(of: " ", with: "").trimmingCharacters(in: .whitespacesAndNewlines)
    guard cleanData.count >= 6 else { return }
    
    telemetryBuffer.withUnsafeMutableBytes { (bufferPointer: UnsafeMutableRawBufferPointer) in
      guard let baseAddress = bufferPointer.baseAddress else { return }
      
      if cleanData.hasPrefix("410C") && cleanData.count >= 8 { // RPM
        let hexVal = String(cleanData.prefix(8).suffix(4))
        if let val = Int(hexVal, radix: 16) {
          let rpm = Int32(val / 4)
          baseAddress.storeBytes(of: rpm.littleEndian, toByteOffset: 0, as: Int32.self)
        }
      } else if cleanData.hasPrefix("410D") && cleanData.count >= 6 { // Speed
        let hexVal = String(cleanData.prefix(6).suffix(2))
        if let val = Int(hexVal, radix: 16) {
          let speed = Int32(val)
          baseAddress.storeBytes(of: speed.littleEndian, toByteOffset: 4, as: Int32.self)
        }
      } else if cleanData.hasPrefix("4105") && cleanData.count >= 6 { // Coolant Temperature
        let hexVal = String(cleanData.prefix(6).suffix(2))
        if let val = Int(hexVal, radix: 16) {
          let coolant = Int32(val - 40)
          baseAddress.storeBytes(of: coolant.littleEndian, toByteOffset: 8, as: Int32.self)
        }
      } else if cleanData.hasPrefix("4111") && cleanData.count >= 6 { // Throttle Position
        let hexVal = String(cleanData.prefix(6).suffix(2))
        if let val = Int(hexVal, radix: 16) {
          let throttle = Int32((val * 100) / 255)
          baseAddress.storeBytes(of: throttle.littleEndian, toByteOffset: 12, as: Int32.self)
        }
      } else if cleanData.hasPrefix("4142") && cleanData.count >= 8 { // Voltage
        let hexVal = String(cleanData.prefix(8).suffix(4))
        if let val = Int(hexVal, radix: 16) {
          let voltage = Double(val) / 1000.0
          baseAddress.storeBytes(of: voltage.bitPattern.littleEndian, toByteOffset: 16, as: UInt64.self)
        }
      }
    }
  }
}

File: modules/motocortex-obd/android/src/main/java/com/motocortex/obd/MotoCortexOBDModule.kt
package com.motocortex.obd

import com.motocortex.obd.session.OBDSessionManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.nio.ByteBuffer
import kotlinx.coroutines.runBlocking

class MotoCortexOBDModule : Module() {
    private var _sessionManager: OBDSessionManager? = null

    private val sessionManager: OBDSessionManager
        get() {
            if (_sessionManager == null) {
                val ctx = appContext.reactContext ?: throw IllegalStateException("OBD Module: ReactContext not ready")
                val manager = OBDSessionManager(ctx)
                manager.setOnStateChangedListener { state ->
                    sendEvent("onConnectionStateChanged", mapOf("state" to state))
                }
                manager.setTelemetryListener { data ->
                    parseAndMutateTelemetry(data)
                }
                _sessionManager = manager
            }
            return _sessionManager!!
        }

    // Direct ByteBuffer for Zero-Allocation JSI Ring Buffer
    // RPM (Int, 4 bytes) at 0
    // Speed (Int, 4 bytes) at 4
    // Coolant (Int, 4 bytes) at 8
    // Throttle (Int, 4 bytes) at 12
    // Voltage (Double, 8 bytes) at 16
    private val telemetryBuffer: ByteBuffer = ByteBuffer.allocateDirect(64)

    override fun definition() = ModuleDefinition {
        Name("MotoCortexOBDModule")

        Events("onConnectionStateChanged")

        OnCreate {
            // Deferred initialization to avoid accessing ReactContext during initial module registration
        }

        AsyncFunction("connectDeviceAsync") { type: String, target: String ->
            sessionManager.connect(type, target)
        }

        AsyncFunction("disconnectDeviceAsync") {
            sessionManager.disconnect()
        }

        AsyncFunction("writeCommandAsync") { command: String ->
            val q = sessionManager.queue ?: throw IllegalStateException("OBD Module: Device not connected")
            runBlocking {
                q.enqueue(command, 1000L).await()
            }
        }

        Function("abortPendingCommands") {
            sessionManager.queue?.abortPendingCommands()
        }

        /**
         * Returns a direct reference to the ByteBuffer (maps to ArrayBuffer in JS).
         * Bypasses event bridging to support 20 Hz updates without allocation overhead.
         */
        Function("getTelemetryBuffer") {
            telemetryBuffer
        }
    }

    private fun parseAndMutateTelemetry(data: String) {
        try {
            val cleanData = data.replace(" ", "").trim()
            if (cleanData.startsWith("410C")) { // Engine RPM
                val hexVal = cleanData.substring(4, 8)
                val rpm = (hexVal.toInt(16)) / 4
                telemetryBuffer.putInt(0, rpm)
            } else if (cleanData.startsWith("410D")) { // Vehicle Speed
                val hexVal = cleanData.substring(4, 6)
                val speed = hexVal.toInt(16)
                telemetryBuffer.putInt(4, speed)
            } else if (cleanData.startsWith("4105")) { // Coolant Temperature
                val hexVal = cleanData.substring(4, 6)
                val coolant = hexVal.toInt(16) - 40
                telemetryBuffer.putInt(8, coolant)
            } else if (cleanData.startsWith("4111")) { // Throttle Position
                val hexVal = cleanData.substring(4, 6)
                val throttle = (hexVal.toInt(16) * 100) / 255
                telemetryBuffer.putInt(12, throttle)
            } else if (cleanData.startsWith("4142")) { // Voltage
                val hexVal = cleanData.substring(4, 8)
                val voltage = hexVal.toInt(16) / 1000.0
                telemetryBuffer.putDouble(16, voltage)
            }
        } catch (e: Exception) {
            // Silently ignore malformed frames on noisy raw streams
        }
    }
}
